README.md
Eu is a wrapper around [request](https://github.com/mikeal/request) which can cache
`HTTP GET` requests based on response headers - just like a browser does.

It is used to lower latency when consuming data from HTTP/REST APIs that have proper caching
headers.

"Eu" means "gotten" in French.

This library is heavily inspired from Kevin Swiber's [request-caching](https://github.com/kevinswiber/request-caching).

## Features

* Multiple cache stores:
  * Redis (built-in)
  * In-memory based on [LRU](https://github.com/isaacs/node-lru-cache) (built-in)
  * Medea (3rd party [eu-medea-store](https://github.com/medea/eu-medea-store) plugin by Kevin Swiber)
* Supports both [public and private caching](http://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.9.1).
* Takes advantage of the `ETag` response header by using the `If-None-Match` request header.
* Takes advantage of the `Last-Modified` response header by using the `If-Modified-Since` request header.
* Cache TTL is based on the `Expires` response header or `max-age` value in `Cache-Control`, but can be overridden.
* Highly customisable with sensible defaults.

## Usage

Usage is similar to `request.get`. The main difference is that you have
to create an `Eu` instance:

```javascript
var Eu = require('eu');
var eu = new Eu(cache); // See below for details about how to create a cache

eu.get('http://some.url', function(err, res, body) {
});

// Or pass in request options:
eu.get('http://some.url', {json: true}, function(err, res, body) {
});

```

A cache uses a _store_ to store cached responses:

### In-memory store

```javascript
var LRU = require('lru-cache');
var store = new Eu.MemoryStore(new LRU());
var cache = new Eu.Cache(store);
```

### Redis store

```javascript
var redis = require('redis').createClient();
var store = new Eu.RedisStore(redis);
var cache = new Eu.Cache(store);
```

You can also create a `NullCache`, which does nothing and doesn't require a store:

### Null cache

```javascript
var cache = new Eu.NullCache();
```

### Cache key name spacing

You should always provide a `prefix` which prefixes the key with the name of
your app (or API). When you invoke `Cache.flush`, it will *only* flush the keys
starting with that prefix. If you don't specify a prefix, you'll flush the *entire* cache.

```javascript
var prefix = 'yourapp:';
var cache = new Eu.Cache(store, prefix);

cache.flush(cb); // only the 'yourapp:*' keys get flushed
```

### Private caching

Some HTTP responses should be cached privately - i.e. they shouldn't be available for other users.
This is the case when the server responds with `Cache-Control: private`.

To handle this you should construct the `Cache` with a `privateSuffix` String argument.
This suffix will be appended to the key when caching a private response.

```javascript
var prefix = 'yourapp:';
var unique = req.user.id; // or req.cookies['connect.sid']
var privateSuffix = ':private:' + unique;
var cache = new Eu.Cache(store, prefix, privateSuffix);
```

You will get an error if no `privateSuffix` was provided when caching a private response.

For servers that don't reply with a `Cache-Control: private` header, you can force the request to be cached privately:

```javascript
cache.get(url, { private: true }, function (err, val) {
  ...
});
```

Eu will always look in the public cache first, and then in the private cache if there was no
hit in the public cache.

### Cache entry expiry time

Every time a response is cached, it is cached with an expiry time. This is a timestamp indicating
how long the cache entry is valid for. (This is not the same as TTL - Time to Live. See next section).

When Eu finds an entry in the cache it will consult this timestamp to decide whether the
entry is still valid - i.e. is the expiry time in the future.

If it's valid, the response is returned immediately. If not, a HTTP request will be issued,
setting the `If-None-Match` request header to the value of the cached `ETag` value,
and the `If-Modified-Since` request header to the value of the cached `Last-Modified` value.

If the server responds with `304 Not Modified` the cached response will be returned even
though it is expired (the server has confirmed that even though it is expired, it hasn't changed).

By default the expiry time is determined by either the `Cache-Control max-age` or `Expires` response
header. If neither of these headers are set, or if you want to cache more aggressively than
what they indicate, you can override this in the options passed to `eu.get`:

```javascript
function expiryTimeMillis() {
  return 8640000000000000; // cache forever
}
cache.get(url, { expiryTimeMillis: expiryTimeMillis }, function (err, val) {
  ...
});
```

### TTL

By default, Eu will store entries in the cache with the TTL (Time to Live) equal to the cache
expiry time (see above).

If the expiry time is undefined, the response will be cached forever, or until the cache discards it, using
whatever expiry algorithm it uses, such as LIFO or FIFO.

You can override the default TTL by supplying your own `ttl` function:

```javascript
function myTtl(ttl) {
  return ttl * 1000; // cache it longer than the server told us to.
}

var cache = new Eu.Cache(store, null, null, myTtl);
```

## Debugging

Just set `DEBUG=eu` in your shell to see debug info.

For extra verbose output (print request and response headers) set `DEBUG=eu,eu:*`


History.md
# [v1.3.2](https://github.com/cucumber-ltd/eu/compare/v1.3.1...v1.3.2) (2015-03-05)

* Fix splitting of cache-control header value ([#6](https://github.com/cucumber-ltd/eu/pull/6))

# [v1.3.1](https://github.com/cucumber-ltd/eu/compare/v1.2.2...v1.3.1) (2015-02-03)

* Better debugging of request/response headers
* Allow overriding of cache entry expiry time
* Link to [eu-medea-store](https://github.com/medea/eu-medea-store) ([#4](https://github.com/cucumber-ltd/eu/issues/4))

# [v1.2.2](https://github.com/cucumber-ltd/eu/compare/v1.2.1...v1.2.2) (2014-07-08)

* Update deps

# [v1.2.1](https://github.com/cucumber-ltd/eu/compare/v1.2.0...v1.2.1) (2014-07-08)

* Add debugging
* Fix caching bug: only set expiry date when explicitly sent by the server

# [v1.2.0](https://github.com/cucumber-ltd/eu/compare/v1.1.1...v1.2.0) (2014-07-08)

* Add option to force private caching

# [v1.1.1](https://github.com/cucumber-ltd/eu/compare/v1.1.0...v1.1.1) (2014-05-30)

* Make sure "Cache-Control: no-cache" isn't cached ([#1](https://github.com/cucumber-ltd/eu/pull/1) Jari Bakken)
* Make it possible to run tests without redis using `NO_REDIS=1 npm test` ([#2](https://github.com/cucumber-ltd/eu/pull/2) Jari Bakken)

# 1.1.0 (Cucumber Limited internal release)


test/eu_test.js
var assert = require('assert');
var http = require('http');
var Eu = require('..');
var stores = require('./helpers/test_stores');

var port = 9090;

stores.forEach(function (store) {
  var paulsCache = new Eu.Cache(store, 'prefix:', ':private:paul');
  var lisasCache = new Eu.Cache(store, 'prefix:', ':private:lisa');

  describe('Eu with ' + store.constructor.name, function () {
    beforeEach(store.flushAll);

    it('caches publicly for Cache-Control: max-age=300', function (cb) {
      http.createServer(function (req, res) {
        var date = new Date().toUTCString();
        res.writeHead(200, { 'Date': date, 'Cache-Control': 'max-age=300' });
        res.end('Cachifiable!');
      }).listen(++port, function () {
        var eu = new Eu(paulsCache);
        eu.get('http://localhost:' + port, function (err, res) {
          if (err) return cb(err);
          lisasCache.get('http://localhost:' + port, function (err, val) {
            if (err) return cb(err);
            assert.equal(val.response.body, 'Cachifiable!');
            cb();
          });
        });
      });
    });

    it('caches privately for Cache-Control: private, max-age=300', function (cb) {
      http.createServer(function (req, res) {
        var date = new Date().toUTCString();
        res.writeHead(200, { 'Date': date, 'Cache-Control': 'private, max-age=300' });
        res.end('Cachifiable!');
      }).listen(++port, function () {
        var eu = new Eu(paulsCache);
        eu.get('http://localhost:' + port, {}, function (err, res) {
          if (err) return cb(err);
          paulsCache.get('http://localhost:' + port, function (err, val) {
            if (err) return cb(err);
            assert.equal(val.response.body, 'Cachifiable!');
            lisasCache.get('http://localhost:' + port, function (err, val) {
              if (err) return cb(err);
              assert.equal(val, undefined);
              cb();
            });
          });
        });
      });
    });

    it('caches privately with option private: true', function (cb) {
      http.createServer(function (req, res) {
        var date = new Date().toUTCString();
        res.writeHead(200, { 'Date': date, 'Cache-Control': 'max-age=300' });
        res.end('Cachifiable!');
      }).listen(++port, function () {
        var eu = new Eu(paulsCache);
        eu.get('http://localhost:' + port, { private: true }, function (err, res) {
          if (err) return cb(err);
          lisasCache.get('http://localhost:' + port, function (err, val) {
            if (err) return cb(err);
            assert.equal(val, undefined);
            cb();
          });
        });
      });
    });

    it('serves from cache when within max-age', function (cb) {
      var s = http.createServer(function (req, res) {
        var date = new Date();
        res.writeHead(200, { 'Date': date.toUTCString(), 'Cache-Control': 'max-age=300' });
        res.end('Cachifiable!');
      }).listen(++port, function () {
        var eu = new Eu(paulsCache);
        eu.get('http://localhost:' + port, function (err, res) {
          if (err) return cb(err);
          s.close(function (err) {
            if (err) return cb(err);
            eu.get('http://localhost:' + port, {}, function (err, res, body) {
              if (err) return cb(err);
              assert.equal(body, 'Cachifiable!');
              cb();
            });
          });
        });
      });
    });

    it('caches when Expires header is set', function (cb) {
      http.createServer(function (req, res) {
        var date = new Date();
        var expires = new Date(date.getSeconds() + 30);
        res.writeHead(200, { 'Date': date.toUTCString(), 'Expires': expires.toUTCString() });
        res.end('Cachifiable!');
      }).listen(++port, function () {
        var eu = new Eu(paulsCache);
        eu.get('http://localhost:' + port, {}, function (err, res) {
          if (err) return cb(err);
          paulsCache.get('http://localhost:' + port, function (err, val) {
            if (err) return cb(err);
            assert.equal(val.response.body, 'Cachifiable!');
            cb();
          });
        });
      });
    });

    it("doesn't cache Cache-Control: no-cache", function (cb) {
      http.createServer(function (req, res) {
        var date = new Date();
        res.writeHead(200, { 'Date': date.toUTCString(), 'Cache-Control': 'no-cache' });
        res.end('Not Cachifiable!');
      }).listen(++port, function () {
        var eu = new Eu(paulsCache);
        eu.get('http://localhost:' + port, {}, function (err, res) {
          if (err) return cb(err);
          paulsCache.get('http://localhost:' + port, function (err, val) {
            if (err) return cb(err);
            assert.equal(val, null);
            cb();
          });
        });
      });
    });

    it("doesn't cache when response code is not 2xx", function (cb) {
      http.createServer(function (req, res) {
        var date = new Date();
        var expires = new Date(date.getSeconds() + 30);
        res.writeHead(404, { 'Date': date.toUTCString(), 'Expires': expires.toUTCString() });
        res.end('Not Cachifiable!');
      }).listen(++port, function () {
        var eu = new Eu(paulsCache);
        eu.get('http://localhost:' + port, {}, function (err, res) {
          if (err) return cb(err);
          paulsCache.get('http://localhost:' + port, function (err, val) {
            if (err) return cb(err);
            assert.equal(val, null);
            cb();
          });
        });
      });
    });

    it('re-requests with If-None-Match when Etag is in response', function (cb) {
      var three_o_four = false;
      http.createServer(function (req, res) {
        if (req.headers['if-none-match'] == 'the-etag') {
          three_o_four = true;
          res.writeHead(304);
          res.end();
        } else {
          var date = new Date();
          res.writeHead(200, { 'Date': date.toUTCString(), 'ETag': 'the-etag' });
          res.end('Cachifiable!');
        }
      }).listen(++port, function () {
        var eu = new Eu(paulsCache);
        eu.get('http://localhost:' + port, {}, function (err, res, body) {
          if (err) return cb(err);
          assert.equal(three_o_four, false);
          eu.get('http://localhost:' + port, {}, function (err, res, body) {
            if (err) return cb(err);
            assert(three_o_four);
            assert.equal(body, 'Cachifiable!');
            cb();
          });
        });
      });
    });

    it('re-requests with If-Modified-Since when Last-Modified is in response', function (cb) {
      var last_modified = new Date();
      var three_o_four = false;
      http.createServer(function (req, res) {
        var if_modified_since = req.headers['if-modified-since'] ? new Date(req.headers['if-modified-since']) : null;
        if (if_modified_since && if_modified_since.getTime() <= last_modified.getTime()) {
          three_o_four = true;
          res.writeHead(304);
          res.end();
        } else {
          var date = new Date();
          res.writeHead(200, { 'Date': date.toUTCString(), 'Last-Modified': last_modified.toUTCString() });
          res.end('Cachifiable!');
        }
      }).listen(++port, function () {
        var eu = new Eu(paulsCache);
        eu.get('http://localhost:' + port, {}, function (err, res, body) {
          if (err) return cb(err);
          assert.equal(three_o_four, false);
          eu.get('http://localhost:' + port, {}, function (err, res, body) {
            if (err) return cb(err);
            assert(three_o_four);
            assert.equal(body, 'Cachifiable!');
            cb();
          });
        });
      });
    });
    it('caches privately for Cache-Control: max-age=300, no-transform, private', function (cb) {
      http.createServer(function (req, res) {
        var date = new Date().toUTCString();
        res.writeHead(200, { 'Date': date, 'Cache-Control': 'max-age=300, no-transform, private' });
        res.end('Cachifiable!');
      }).listen(++port, function () {
        var eu = new Eu(paulsCache);
        eu.get('http://localhost:' + port, {}, function (err, res) {
          if (err) return cb(err);
          paulsCache.get('http://localhost:' + port, function (err, val) {
            if (err) return cb(err);
            assert.equal(val.response.body, 'Cachifiable!');
            lisasCache.get('http://localhost:' + port, function (err, val) {
              if (err) return cb(err);
              assert.equal(val, undefined);
              cb();
            });
          });
        });
      });
    });


  });

});


test/cache_test.js
var assert = require('assert');
var Eu = require('..');
var stores = require('./helpers/test_stores');

stores.forEach(function (store) {
  describe('Cache with ' + store.constructor.name, function () {
    beforeEach(store.flushAll);

    it('Keys off uri when no prefix is provided', function (cb) {
      var cache = new Eu.Cache(store);
      cache.set('my-uri', false, 'my-value', 1000, function (err) {
        if (err) return cb(err);
        store.get('my-uri', function (err, value) {
          if (err) return cb(err);
          assert.equal(JSON.parse(value), 'my-value');
          cb();
        });
      });
    });

    it('Requires privateSuffix for private caching', function (cb) {
      var cache = new Eu.Cache(store);
      cache.set('my-uri', true, 'my-value', 1000, function (err) {
        if (!err) return cb(new Error('Should fail'));
        assert.equal(err.message, 'Cannot cache privately without a privateSuffix');
        cb();
      });
    });

    it('Expires cached keys after TTL', function (cb) {
      var cache = new Eu.Cache(store);
      var ttl = 10;
      cache.set('my-uri', false, 'my-value', ttl, function (err) {
        if (err) return cb(err);
        cache.get('my-uri', function (err, value) {
          if (err) return cb(err);
          assert.equal(value, 'my-value');
          setTimeout(function () {
            cache.get('my-uri', function (err, value) {
              if (err) return cb(err);
              assert.equal(value, undefined);
              cb();
            });
          }, ttl*2);
        });
      });
    });

    it('Overrides TTL with custom function', function (cb) {
      function oneMillisecond(ttl) {
        return 1;
      }

      var cache = new Eu.Cache(store, null, null, oneMillisecond);
      cache.set('my-uri', false, 'my-value', 100, function (err) {
        if (err) return cb(err);
        setTimeout(function () {
          cache.get('my-uri', function (err, value) {
            if (err) return cb(err);
            assert.equal(value, undefined);
            cb();
          });
        }, 10);
      });
    });

    it('Deletes Redis keys by match', function (cb) {
      var fooCache = new Eu.Cache(store, 'foo:');
      var barCache = new Eu.Cache(store, 'bar:');

      fooCache.set('KEY', false, 'FOO', 10000, function (err) {
        if (err) return cb(err);
        barCache.set('KEY', false, 'BAR', 10000, function (err) {
          if (err) return cb(err);

          fooCache.flush(function (err, n) {
            if (err) return cb(err);
            assert.equal(n, 1);
            fooCache.get('KEY', function (err, value) {
              if (err) return cb(err);
              assert.equal(value, undefined);
              barCache.get('KEY', function (err, value) {
                if (err) return cb(err);
                assert.equal(value, 'BAR');
                cb();
              });
            });
          });
        });
      });
    });
  });
});


test/helpers/test_stores.js
var Eu = require('../..');

var stores = [];

var LRU = require('lru-cache');
stores.push(new Eu.MemoryStore(new LRU()));

if (!process.env.NO_REDIS) {
  var redis = require('redis').createClient();
  stores.push(new Eu.RedisStore(redis));
}

module.exports = stores;



lib/redis_store.js
/**
 * Creates a new Redis backed store.
 *
 * @param redis a redis client, obtained by require('redis').createClient()
 */
module.exports = function RedisStore(redis) {
  this.set = function (key, value, ttlMillis, cb) {
    if (ttlMillis > 0) {
      redis.psetex(key, ttlMillis, value, cb);
    } else {
      redis.set(key, value, cb);
    }
  };

  this.get = function (key, cb) {
    redis.get(key, cb);
  };

  this.flush = function (prefix, cb) {
    var wildcard = prefix + '*';
    redis.keys(wildcard, function (err, keys) {
      if (err) return cb(err);
      if (keys.length === 0) return cb(null, 0);
      redis.del(keys, function (err) {
        if (err) return cb(err);
        cb(null, keys.length);
      });
    });
  };

  this.flushAll = function (cb) {
    redis.flushdb(cb);
  };
};


lib/memory_store.js
/**
 * Creates a new in-memory LRU backed storage.
 *
 * @param lru an LRU instance, created by var LRU = require('lru'); lru = new LRU();
 */
module.exports = function MemoryStore(lru) {
  this.set = function (key, value, ttlMillis, cb) {
    lru.set(key, value);
    // Max timeout value in JavaScript is 2^31-1
    var timeout = Math.min(0x7FFFFFFF, ttlMillis);
    setTimeout(function () {
      lru.del(key);
    }, timeout);
    cb();
  };

  this.get = function (key, cb) {
    cb(null, lru.get(key));
  };

  this.flush = function (prefix, cb) {
    var deleted = 0;
    lru.forEach(function (value, key) {
      var startsWithPrefix = key.indexOf(prefix) == 0
      if (startsWithPrefix) {
        lru.del(key);
        deleted++;
      }
    })
    cb(null, deleted);
  };

  this.flushAll = function (cb) {
    lru.reset();
    cb();
  }
};


lib/eu.js
var request = require('request');
var debug = require('debug')('eu');
var debugRequest = require('debug')('eu:request');
var END_OF_TIME = 8640000000000000;

if(debugRequest.enabled) {
  function log(type, data) {
    debugRequest(type, data.headers);
  }
  require('request-debug')(request, log);
}

module.exports = function Eu(cache) {
  if(cache === undefined) throw new Error('cache cannot be undefined');

  this.get = function (uri, options, callback) {
    if (typeof uri === 'undefined') throw new Error('undefined is not a valid uri or options object.');
    if (typeof options === 'function' && !callback) {
      callback = options;
      options = {};
    }

    debug('%s get', uri);
    cache.get(uri, function (err, value) {
      if (err) return callback(err);

      debug('%s cache %s', uri, value ? 'hit' : 'miss');

      if (value && new Date().getTime() <= value.expiryTimeMillis) {
        debug('%s cache hit, not expired. No request needed. (expiry: %s)', uri, new Date(value.expiryTimeMillis));
        return callback(null, value.response, value.response.body);
      }

      if (options.headers === undefined) options.headers = {};
      if (options.expiryTimeMillis === undefined) options.expiryTimeMillis = function (expiryTimeMillis) {
        return expiryTimeMillis;
      };

      if (value && 'etag' in value.response.headers) {
        debug('%s setting header If-None-Match = %s', uri, value.response.headers['etag']);
        options.headers['If-None-Match'] = value.response.headers['etag'];
      }
      if (value && 'last-modified' in value.response.headers) {
        debug('%s setting header If-Modified-Since = %s', uri, value.response.headers['last-modified']);
        options.headers['If-Modified-Since'] = value.response.headers['last-modified'];
      }

      request(uri, options, function (err, res, body) {
        if (err) return callback(err, res, body);

        if (res.statusCode == 304) {
          debug('%s 304 Not Modified. Returning cached response.', uri);
          // 304 means it's OK to use the cached value.
          return callback(null, value.response, value.response.body);
        }
        if (res.statusCode >= 300) {
          debug('%s %s (not caching)', uri, res.statusCode);
          // Only cache 2xx responses
          return callback(null, res, body);
        }

        var isCacheable = false;
        var expiryTimeMillis;
        var private = options.private;
        if ('cache-control' in res.headers) {
          // In case of Cache-Control: no-cache, cacheable should remain false.
          var val = res.headers['cache-control'].split(/\s*,\s*/);
          var cacheControl = {};
          val.forEach(function (dir) {
            var arr = dir.split('=');
            if (arr.length == 1) arr.push(true);
            cacheControl[arr[0]] = arr[1];
          });
          private = private || cacheControl.private;
          if (cacheControl['max-age']) {
            isCacheable = true;
            var date = new Date(res.headers['date']);
            var seconds = +cacheControl['max-age'];
            expiryTimeMillis = date.getTime() + 1000 * seconds;
          }
        }
        if ('expires' in res.headers) {
          isCacheable = true;
          var expires = new Date(res.headers['expires']);
          expiryTimeMillis = expires.getTime();
        }
        if ('etag' in res.headers) {
          isCacheable = true;
        }
        if ('last-modified' in res.headers) {
          isCacheable = true;
        }

        if (isCacheable) {
          expiryTimeMillis = options.expiryTimeMillis(expiryTimeMillis);
          var ttlMillis = Math.max((expiryTimeMillis || END_OF_TIME) - Date.now(), 0);
          var cachedResponse = { response: cachedResponseProperties(res), expiryTimeMillis: expiryTimeMillis };

          debug('%s writing %s response to cache with ttl %s ms (%s)', uri, private ? 'private' : 'public', ttlMillis, new Date(ttlMillis + Date.now()));
          cache.set(uri, private, cachedResponse, ttlMillis, function (err) {
            callback(err, res, body);
          });
        } else {
          debug('%s not cacheable - no caching headers in response', uri);
          callback(null, res, body);
        }
      });
    });

    function cachedResponseProperties(res) {
      return {
        statusCode: res.statusCode,
        headers: res.headers,
        body: res.body
      }
    }
  };
};
